src/parse/decl.rs

```diff
@@ -979,6 +981,20 @@ impl<I: Iterator<Item = Lexeme>> Parser<I> {
                 _ => break,
             };
         }
+
+        if let Some(Declarator {
+            current: DeclaratorType::Function(_),
+            ..
+        }) = prefix
+        {
+            // `inline` is allowed on function declarations
+        } else if qualifiers.inline {
+            self.semantic_err(
+                "`inline` is only allowed on function declarations",
+                self.last_location,
+            );
+        }
```

Comment:
So I know I keep changing my mind, but I think this would actually fit better in `Declarator::parse_type`. That way you don't have to keep passing the `qualifiers` down, you can just check them once at the end. Sorry for the inconvenience!

Reply:
Actually given all the other places you have to check maybe it would be better to just have each call site check it individually ...

Reply:
I changed my mind again :laughing:  I think this is fine - it lets us check if we're in a function and the parameter decayed to a pointer. Specifically I'd like to give a better error than clang here:

```c
int f(inline int g());
```

clang: `<stdin>:1:7: error: 'inline' can only appear on functions`
gcc: `<stdin>:1:18: warning: parameter ‘g’ declared ‘inline’`
what I'd like to see: `error: 'inline' can only appear on functions. help: function parameters decay to pointers`

So long story short your current code is fine :)